File: python/streamlines/kde.py
"""
Kernel density estimation.
"""

import pyopencl as cl
import pyopencl.array
import numpy as np
from scipy.stats import norm
import os
os.environ['PYTHONUNBUFFERED']='True'
import warnings

from streamlines import pocl
from streamlines.useful import vprint, pick_seeds

__all__ = ['estimate_univariate_pdf','estimate_bivariate_pdf',
           'gpu_compute_1d','prepare_memory_1d']

pdebug = print

def estimate_bivariate_pdf( distbn ):
        
    """
    Compute bivariate histogram and subsequent kernel-density smoothed pdf.
    
    Args:
        distbn (obj):
    
    Returns:
        
        
    """
    verbose = distbn.verbose
    vprint(verbose,'Estimating bivariate pdf...',end='')
    
    # Prepare CL essentials
    platform,device,context= pocl.prepare_cl_context(distbn.cl_platform,distbn.cl_device)
    queue = cl.CommandQueue(context)
    
    cl_files = ['kde.cl']
    cl_kernel_source = ''
    for cl_file in cl_files:
        with open(os.path.join(distbn.cl_src_path,cl_file), 'r') as fp:
            cl_kernel_source += fp.read()
            
    sl_array     = distbn.logxy_data
    n_data       = distbn.n_data
    bandwidth    = distbn.bandwidth
    
    x_range      = distbn.logx_range
    bin_dx       = distbn.bin_dx
    pdf_dx       = distbn.pdf_dx
    stddev_x     = np.std(sl_array[:,0])
#     pdebug('\nstd x={0:0.2} lxmin={1:0.2}  lxmax={2:0.2}=>{3:0.0f}'
#            .format(stddev_x,  distbn.logx_min,distbn.logx_max,
#                    np.exp(distbn.logx_max)))
    
    y_range      = distbn.logy_range
    bin_dy       = distbn.bin_dy
    pdf_dy       = distbn.pdf_dy
    stddev_y     = np.std(sl_array[:,1])
#     pdebug('std y={0:0.2} lymin={1:0.2}  lymax={2:0.2}' 
#            .format(stddev_y,distbn.logy_min,distbn.logy_max))

    # Set up kernel filter
    # Hacked Silverman hack
    distbn.kdf_width_x = 1.06*stddev_x*np.power(n_data,-0.2)*bandwidth
    distbn.kdf_width_y = 1.06*stddev_y*np.power(n_data,-0.2)*bandwidth
    distbn.n_kdf_part_points_x= (np.uint32(np.floor(distbn.kdf_width_x/bin_dx))//2)
    distbn.n_kdf_part_points_y= (np.uint32(np.floor(distbn.kdf_width_y/bin_dy))//2)
            
    vprint(verbose,'histogram...',end='')
    # Histogram
    cl_kernel_fn = 'histogram_bivariate'
    histogram_array \
        = gpu_compute(device, context, queue, cl_kernel_source, cl_kernel_fn, 
                      distbn, action='histogram',  is_bivariate=True, 
                      sl_array=sl_array, verbose=verbose)
        
    vprint(verbose,'kernel filtering rows..',end='')
    # PDF
    cl_kernel_fn = 'pdf_bivariate_rows'
    partial_pdf_array \
        = gpu_compute(device, context, queue, cl_kernel_source, cl_kernel_fn, 
                      distbn, action='partial_pdf',  is_bivariate=True, 
                      histogram_array=histogram_array,
                      verbose=verbose)

    vprint(verbose,'kernel filtering columns...',end='')
    cl_kernel_fn = 'pdf_bivariate_cols'
    pdf_array \
        = gpu_compute(device, context, queue, cl_kernel_source, cl_kernel_fn, 
                      distbn,  action='full_pdf',  is_bivariate=True, 
                      partial_pdf_array=partial_pdf_array,
                      verbose=verbose)
    # Done
    vprint(verbose,'done')
    return (pdf_array/(np.sum(pdf_array)*bin_dx*bin_dy), 
            histogram_array/(np.sum(histogram_array)*bin_dx*bin_dy))
    
def estimate_univariate_pdf( distbn, do_detrend=False, logx_vec=None):
        
    """
    Compute univariate histogram and subsequent kernel-density smoothed pdf.
    
    Args:
        distbn (obj):
        do_detrend (bool):
        logx_vec (numpy.ndarray):
    
    Returns:
        
        
    """
    verbose = distbn.verbose
    vprint(verbose,'Estimating univariate pdf...',end='')
    
    # Prepare CL essentials
    platform,device,context= pocl.prepare_cl_context(distbn.cl_platform,distbn.cl_device)
    queue = cl.CommandQueue(context)
    
    cl_files = ['kde.cl']
    cl_kernel_source = ''
    for cl_file in cl_files:
        with open(os.path.join(distbn.cl_src_path,cl_file), 'r') as fp:
            cl_kernel_source += fp.read()
            
    sl_array     = distbn.logx_data
    n_data       = distbn.n_data
    x_range      = distbn.logx_range
    bandwidth    = distbn.bandwidth
    bin_dx       = distbn.bin_dx
    pdf_dx       = distbn.pdf_dx
    stddev_x     = np.std(sl_array)
    
    # Set up kernel filter
    # Hacked Silverman hack
    distbn.kdf_width_x = 1.06*stddev_x*np.power(n_data,-0.2)*bandwidth
    distbn.kdf_width_y = 0.0
    distbn.n_kdf_part_points_x = np.uint32(np.floor(distbn.kdf_width_x/pdf_dx))//2
    distbn.n_kdf_part_points_y = 0
    
#     if not do_detrend:
#         pdebug('\nstd x={0:0.2} lxmin={1:0.2}  lxmax={2:0.2}'
#            .format(stddev_x,  distbn.logx_min,distbn.logx_max))
        
    vprint(verbose,'histogram...',end='')
    # Histogram
    cl_kernel_fn = 'histogram_univariate'
    histogram_array \
        = gpu_compute(device, context, queue, cl_kernel_source, cl_kernel_fn, 
                      distbn, action='histogram', sl_array=sl_array, 
                      verbose=verbose)
        
    if do_detrend:
        x = logx_vec
        pdf = histogram_array.copy().astype(np.float32)
        mean = (np.sum(x*pdf)/np.sum(pdf))
        stddev = np.sqrt(np.sum( (x-mean)**2 * pdf)/np.sum(pdf))
#         pdebug(np.exp(mean),np.exp(stddev))
        norm_pdf = norm.pdf(logx_vec,mean,stddev)
        pdf /= norm_pdf
        pdf /= np.sum(pdf)
        pdf *= np.sum(histogram_array)
        histogram_array = pdf.copy().astype(np.uint32)
        
    vprint(verbose,'kernel filtering..',end='')
    # PDF
    cl_kernel_fn = 'pdf_univariate'
    pdf_array \
        = gpu_compute(device, context, queue, cl_kernel_source, cl_kernel_fn, 
                      distbn, action='full_pdf', histogram_array=histogram_array,
                      verbose=verbose)
    # Done
    vprint(verbose,'done')
    
    return (pdf_array/(sum(pdf_array)*bin_dx),
            histogram_array/(sum(histogram_array)*bin_dx))
    
def gpu_compute( device, context, queue, cl_kernel_source, cl_kernel_fn, distbn,
                 action='histogram', is_bivariate=False, 
                 sl_array=None,  histogram_array=None, 
                 partial_pdf_array=None,  pdf_array=None, 
                 verbose=False ):
    """
    Carry out GPU computation of histogram.
    
    Args:
        device (pyopencl.Device):
        context (pyopencl.Context):
        queue (pyopencl.CommandQueue):
        cl_kernel_source (str):
        cl_kernel_fn (str):
        distbn (obj):
        sl_array (numpy.ndarray):
        histogram_array (numpy.ndarray):
        verbose (bool):  
        
    """
        
    # Prepare memory, buffers 
    n_hist_bins  = distbn.n_hist_bins
    n_data       = distbn.n_data
    n_pdf_points = distbn.n_pdf_points
        
    if action=='histogram':
        # Compute histogram
        (histogram_array, sl_buffer, histogram_buffer) \
            = prepare_memory(context, queue, 
                             n_hist_bins=n_hist_bins, 
                             sl_array=sl_array, 
                             action=action, is_bivariate=is_bivariate,
                             verbose=verbose)    
        global_size = [n_data,1]
        buffer_list = [sl_buffer, histogram_buffer]
    elif action=='partial_pdf':
        # Compute partially (rows-only) kd-smoothed pdf
        (partial_pdf_array, histogram_buffer, partial_pdf_buffer) \
            = prepare_memory(context, queue, 
                             n_hist_bins=n_hist_bins,
                             histogram_array=histogram_array,
                             action=action, is_bivariate=is_bivariate,
                             verbose=verbose)    
        global_size = [partial_pdf_array.shape[0]*partial_pdf_array.shape[1],1]
        buffer_list = [histogram_buffer, partial_pdf_buffer]
    else:
        # Compute kd-smoothed pdf
        if is_bivariate:
            (pdf_array, partial_pdf_buffer, pdf_buffer) \
                = prepare_memory(context, queue, 
                                 n_pdf_points=n_pdf_points, 
                                 partial_pdf_array=partial_pdf_array, 
                                 action=action, is_bivariate=True,
                                 verbose=verbose)    
            global_size = [pdf_array.shape[0]*pdf_array.shape[1],1]
            buffer_list = [partial_pdf_buffer, pdf_buffer]
        else:
            (pdf_array, histogram_buffer, pdf_buffer) \
                = prepare_memory(context, queue, 
                                 n_pdf_points=n_pdf_points, 
                                 histogram_array=histogram_array,
                                 action=action, is_bivariate=False,
                                 verbose=verbose)    
            global_size = [pdf_array.shape[0],1]
            buffer_list = [histogram_buffer, pdf_buffer]
        
    local_size = None
    # Compile the CL code
    compile_options = pocl.set_compile_options(distbn, cl_kernel_fn, job_type='kde')
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        program = cl.Program(context, cl_kernel_source).build(options=compile_options)
    pocl.report_build_log(program, device, verbose)
    # Set the GPU kernel
    kernel = getattr(program,cl_kernel_fn)
    # Designate buffered arrays
    kernel.set_args(*buffer_list)
    kernel.set_scalar_arg_dtypes( [None]*len(buffer_list) )
    # Do the GPU compute
    event = cl.enqueue_nd_range_kernel(queue, kernel, global_size, local_size)
    
    # Fetch the data back from the GPU and finish
    if action=='histogram':
        # Compute histogram
        cl.enqueue_copy(queue, histogram_array, histogram_buffer)
        queue.finish()   
        return histogram_array
    elif action=='partial_pdf':
        # Compute partial pdf
        cl.enqueue_copy(queue, partial_pdf_array, partial_pdf_buffer)
        queue.finish()
        return partial_pdf_array
    else:
        # Compute pdf
        cl.enqueue_copy(queue, pdf_array, pdf_buffer)
        queue.finish()
        return pdf_array
    
def prepare_memory(context, queue, 
                   action='histogram', is_bivariate=False, 
                   n_hist_bins=0, n_pdf_points=0,
                   sl_array=None, histogram_array=None, 
                   partial_pdf_array=None, pdf_array=None, 
                   verbose=False ):
    """
    Create PyOpenCL buffers and np-workalike arrays to allow CPU-GPU data transfer.
    
    Args:
        context (pyopencl.Context):
        queue (pyopencl.CommandQueue):
        n_bins (int):
        sl_array (numpy.ndarray):
        histogram_array (numpy.ndarray):
        pdf_array (numpy.ndarray):
        verbose (bool):
        
    Returns:
        numpy.ndarray, pyopencl.Buffer, pyopencl.Buffer: 
        histogram_array or pdf_array, sl_buffer or histogram_buffer, 
        histogram_buffer or pdf_buffer
    """
    COPY_READ_ONLY  = cl.mem_flags.READ_ONLY  | cl.mem_flags.COPY_HOST_PTR
    COPY_READ_WRITE = cl.mem_flags.READ_WRITE | cl.mem_flags.COPY_HOST_PTR

    if action=='histogram':
        nx = n_hist_bins
        if is_bivariate:
            ny = n_hist_bins
        else:
            ny = 1
        sl_buffer         = cl.Buffer(context, COPY_READ_ONLY,  hostbuf=sl_array)
        histogram_array   = np.zeros((nx,ny), dtype=np.uint32)
        histogram_buffer  = cl.Buffer(context, COPY_READ_WRITE, hostbuf=histogram_array)
        return (histogram_array, sl_buffer, histogram_buffer) 
    elif action=='partial_pdf':
        nx = n_hist_bins
        ny = n_hist_bins
        histogram_buffer  = cl.Buffer(context, COPY_READ_ONLY, hostbuf=histogram_array)
        partial_pdf_array = np.zeros((nx,ny), dtype=np.float32)
        partial_pdf_buffer= cl.Buffer(context, COPY_READ_WRITE,hostbuf=partial_pdf_array)
        return (partial_pdf_array, histogram_buffer, partial_pdf_buffer) 
    else:
        nx = n_pdf_points
        if is_bivariate:
            ny = n_pdf_points
            partial_pdf_buffer=cl.Buffer(context,COPY_READ_ONLY,hostbuf=partial_pdf_array)
            pdf_array         = np.zeros((nx,ny), dtype=np.float32)
            pdf_buffer        = cl.Buffer(context, COPY_READ_WRITE, hostbuf=pdf_array)
            return (pdf_array, partial_pdf_buffer, pdf_buffer) 
        else:
            histogram_buffer = cl.Buffer(context, COPY_READ_ONLY, hostbuf=histogram_array)
            pdf_array         = np.zeros((nx,1), dtype=np.float32)
            pdf_buffer        = cl.Buffer(context, COPY_READ_WRITE, hostbuf=pdf_array)
            return (pdf_array, histogram_buffer, pdf_buffer) 
        

File: ocaml/streamlines/kde.ml
(** {v Copyright (C) 2017-2018,  Colin P Stark and Gavin J Stark.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file   kde.ml
 * @brief  Kernel density estimation
 *
 * Up to date with python of git CS 189bfccdabc3371eafe8bcafa3bdfa8c241e56e4
 * Except  max_time_per_kernel and initial_size_factor are hardwired
 * except for bivariate which is not there yet
 *
 * v}
 *)

(*f  Module abbreviations *)
open Globals
open Properties
open Core
module ODM = Owl.Dense.Matrix.Generic
module ODN = Owl.Dense.Ndarray.Generic
module BA=Bigarray

(** {1 Types } *)

(**  [t]

  Structure for the Preprocess workflow

 *)
type t = {
    props : t_props_preprocess;
    mutable roi_gradx_array : t_ba_floats;
    mutable roi_grady_array : t_ba_floats;
    mutable where_looped : (int * int) list;
    pad_width : int;
  }

(** {1 pv_verbosity functions} *)

(**  [pv_noisy t]

  Shortcut to use {!type:Properties.t_props_trace} verbosity for {!val:Properties.pv_noisy}

 *)
let pv_noisy   (props:t_props_analysis) = Workflow.pv_noisy props.workflow

(**  [pv_debug t]

  Shortcut to use {!type:Properties.t_props_trace} verbosity for {!val:Properties.pv_debug}

 *)
let pv_debug   (props:t_props_analysis) = Workflow.pv_noisy props.workflow

(**  [pv_info t]

  Shortcut to use {!type:Properties.t_props_trace} verbosity for {!val:Properties.pv_info}

 *)
let pv_info    (props:t_props_analysis) = Workflow.pv_info props.workflow

(**  [pv_verbose t]

  Shortcut to use {!type:Properties.t_props_trace} verbosity for {!val:Properties.pv_verbose}

 *)
let pv_verbose (props:t_props_analysis) = Workflow.pv_verbose props.workflow

(** {1 Memory functions} *)

(**  [copy_read pocl]

  Shortcut for creating an OpenCL buffer that is read-only by the
  kernel and copied from a big array before execution

  @return OpenCL buffer

 *)
let copy_read       pocl = Pocl.buffer_of_array pocl ~copy:true true false

(**  [copy_read_write pocl]

  Shortcut for creating an OpenCL buffer that is read-write by the
  kernel and copied from a big array before execution

  @return OpenCL buffer

 *)
let copy_read_write pocl = Pocl.buffer_of_array pocl ~copy:true true true

(**  [write_only pocl]

  Shortcut for creating an OpenCL buffer that is write-only by the
  kernel

  @return OpenCL buffer

 *)
let write_only      pocl = Pocl.buffer_of_array pocl false true

(** {1 GPU compute functions} *)

(**  [gpu_compute_histogram pocl info program sl_array is_bivariate]
 *)
let gpu_compute_histogram pocl info program sl_array is_bivariate =
  let n_hist_bins  = Info.int_of info "n_hist_bins" in
  let n_data       = Info.int_of info "n_data" in

  let cl_kernel_name = if is_bivariate then "histogram_bivariate" else  "histogram_univariate" in
  let kernel = Pocl.get_kernel pocl program cl_kernel_name in

  let nx = n_hist_bins in
  let ny = if is_bivariate then n_hist_bins else 1 in
  let histogram_array    = ba_int2d nx ny in
  ODM.fill histogram_array 0;
  let sl_buffer          = copy_read       pocl sl_array in
  let histogram_buffer   = copy_read_write pocl histogram_array in
  Pocl.kernel_set_arg_buffer pocl kernel 0 sl_buffer;
  Pocl.kernel_set_arg_buffer pocl kernel 1 histogram_buffer;

  let time_taken = Pocl.adaptive_enqueue_kernel pocl kernel n_data 64 in
  Printf.printf "\n##### Kernel lapsed time: %0.3f secs #####\n" time_taken;

  Pocl.copy_buffer_from_gpu pocl ~src:histogram_buffer    ~dst:histogram_array;
  Pocl.finish_queue pocl;
  histogram_array

(**  [gpu_compute_partial_pdf pocl info program histogram_array]
 *)
let gpu_compute_partial_pdf pocl info program histogram_array =
  let n_hist_bins  = Info.int_of info "n_hist_bins" in

  let cl_kernel_name = "pdf_bivariate_rows" in
  let kernel = Pocl.get_kernel pocl program cl_kernel_name in

  let partial_pdf_array  = ba_float2d n_hist_bins n_hist_bins in
  ODM.fill partial_pdf_array 0.;
  let histogram_buffer   = copy_read       pocl histogram_array in
  let partial_pdf_buffer = copy_read_write pocl partial_pdf_array in

  Pocl.kernel_set_arg_buffer pocl kernel 0 histogram_buffer;
  Pocl.kernel_set_arg_buffer pocl kernel 1 partial_pdf_buffer;

  let global_size = n_hist_bins*n_hist_bins in
  let time_taken = Pocl.adaptive_enqueue_kernel pocl kernel global_size 64 in
  Printf.printf "\n##### Kernel lapsed time: %0.3f secs #####\n" time_taken;
  Pocl.copy_buffer_from_gpu pocl ~src:partial_pdf_buffer    ~dst:partial_pdf_array;
  Pocl.finish_queue pocl;
  partial_pdf_array

(**  [gpu_compute_full_bivariate_pdf pocl info program partial_pdf_array]
 *)
let gpu_compute_full_bivariate_pdf pocl info program partial_pdf_array =
  let n_pdf_points  = Info.int_of info "n_pdf_points" in

  let cl_kernel_name = "pdf_bivariate_cols" in
  let kernel = Pocl.get_kernel pocl program cl_kernel_name in

  let pdf_array          = ba_float2d n_pdf_points n_pdf_points in
  ODM.fill pdf_array 0.;
  let partial_pdf_buffer = copy_read       pocl partial_pdf_array in
  let pdf_buffer         = copy_read_write pocl pdf_array in

  Pocl.kernel_set_arg_buffer pocl kernel 0 partial_pdf_buffer;
  Pocl.kernel_set_arg_buffer pocl kernel 1 pdf_buffer;

  let global_size = n_pdf_points*n_pdf_points in
  let time_taken = Pocl.adaptive_enqueue_kernel pocl kernel global_size 64 in
  Printf.printf "\n##### Kernel lapsed time: %0.3f secs #####\n" time_taken;
  Pocl.copy_buffer_from_gpu pocl ~src:pdf_buffer    ~dst:pdf_array;
  Pocl.finish_queue pocl;
  pdf_array

(**  [gpu_compute_full_univariate_pdf pocl info program histogram_array]
 *)
let gpu_compute_full_univariate_pdf pocl info program histogram_array =
  let n_pdf_points  = Info.int_of info "n_pdf_points" in

  let cl_kernel_name = "pdf_univariate" in
  let kernel = Pocl.get_kernel pocl program cl_kernel_name in

  let pdf_array          = ba_float2d n_pdf_points 1 in
  ODM.fill pdf_array 0.;
  let histogram_buffer   = copy_read       pocl histogram_array in
  let pdf_buffer         = copy_read_write pocl pdf_array in

  Pocl.kernel_set_arg_buffer pocl kernel 0 histogram_buffer;
  Pocl.kernel_set_arg_buffer pocl kernel 1 pdf_buffer;

  let global_size = n_pdf_points in
  let time_taken = Pocl.adaptive_enqueue_kernel pocl kernel global_size 64 in
  Printf.printf "\n##### Kernel lapsed time: %0.3f secs #####\n" time_taken;
  Pocl.copy_buffer_from_gpu pocl ~src:pdf_buffer    ~dst:pdf_array;
  Pocl.finish_queue pocl;
  pdf_array

(** {1 Statics} *)

let cl_files = ["kde.cl";
    ]
let cl_src_path = ["opencl"]

(** {1 PDF estimation }
 *)

(**  [estimate_univariate_pdf t pocl info sl_array]

For KDE we need to set
            '-D','KERNEL_{}'.format(kernel_def.upper()),
            '-D','KDF_BANDWIDTH={}f'.format(info_dict['kdf_bandwidth']),
            '-D','KDF_IS_{}'.format(info_dict['kdf_kernel'].upper()),
            '-D','N_DATA={}u'.format(info_dict['n_data']),
            '-D','N_HIST_BINS={}u'.format(info_dict['n_hist_bins']),
            '-D','N_PDF_POINTS={}u'.format(info_dict['n_pdf_points']),
            '-D','X_MIN={}f'.format(info_dict['x_min']),
            '-D','X_MAX={}f'.format(info_dict['x_max']),
            '-D','X_RANGE={}f'.format(info_dict['x_range']),
            '-D','BIN_DX={}f'.format(info_dict['bin_dx']),
            '-D','PDF_DX={}f'.format(info_dict['pdf_dx']),
            '-D','KDF_WIDTH_X={}f'.format(info_dict['kdf_width_x']),
            '-D','N_KDF_PART_POINTS_X={}u'.format(info_dict['n_kdf_part_points_x']),
            '-D','Y_MIN={}f'.format(info_dict['y_min']),
            '-D','Y_MAX={}f'.format(info_dict['y_max']),
            '-D','Y_RANGE={}f'.format(info_dict['y_range']),
            '-D','BIN_DY={}f'.format(info_dict['bin_dy']),
            '-D','PDF_DY={}f'.format(info_dict['pdf_dy']),
            '-D','KDF_WIDTH_Y={}f'.format(info_dict['kdf_width_y']),
            '-D','N_KDF_PART_POINTS_Y={}u'.format(info_dict['n_kdf_part_points_y'])

 *)
let estimate_univariate_pdf t pocl info sl_array =

  let cl_kernel_source = Pocl.read_source cl_src_path cl_files in
            
  let n_data   = Info.int_of   info "n_data" in
  let bin_dx   = Info.float_of info "bin_dx" in
  let pdf_dx   = Info.float_of info "pdf_dx" in

  let stddev=1.0 in (*    stddev       = np.std(sl_array)*)
    
  (* Set up kernel filter -  Silverman hack for now *)
  let kdf_width_x = 1.06 *. stddev *. ((float n_data) ** (-0.2)) *. 8. in
  let n_kdf_part_points_x = (truncate (kdf_width_x /. pdf_dx)) / 2 in
  Info.set info "kdf_width_x"         (Info.Float32 kdf_width_x);
  Info.set info "n_kdf_part_points_x" (Info.Int     n_kdf_part_points_x);

  Info.set_int info "kernel_histogram_univariate" 1;
  Info.set_int info "kernel_pdf_univariate" 1;
  Info.set_int info (sfmt "kdf_is_%s" (Info.str_of info "kernel")) 1;

  let compile_options = Pocl.compile_options pocl info "KDE" in
  let program = Pocl.compile_program pocl cl_kernel_source compile_options in

  pv_verbose t (fun _ -> Printf.printf "histogram...%!");
  let histogram_array = gpu_compute_histogram pocl info program sl_array false in
  pv_verbose t (fun _ -> Printf.printf "done\n%!");

  pv_verbose t (fun _ -> Printf.printf "kernel filtering...%!");
  let pdf_array = gpu_compute_full_univariate_pdf pocl info program histogram_array in
  pv_verbose t (fun _ -> Printf.printf "done\n%!");

  let weighted_total = (ODM.sum' pdf_array) *. bin_dx in
  ODM.div_scalar pdf_array weighted_total

(**  [estimate_bivariate_pdf pocl data]

    Compute bivariate histogram and subsequent kernel-density smoothed pdf.

 *)
let estimate_bivariate_pdf t pocl info sl_array =

  let cl_kernel_source = Pocl.read_source cl_src_path cl_files in
            
  let bandwidth   = Info.float_of info  "kdf_bandwidth" in

  let x_range  = Info.float_of info "x_range" in
  let stddev_x  = 1.0 in
  let bin_dx   = Info.float_of info "bin_dx" in
  let pdf_dx   = Info.float_of info "pdf_dx" in

  let y_range  = Info.float_of info "y_range" in
  let stddev_y  = 1.0 in
  let bin_dy   = Info.float_of info "bin_dy" in
  let pdf_dy   = Info.float_of info "pdf_dy" in

  (* Set up kernel filter -  Silverman hack for now *)
  let kdf_width_x = stddev_x *. bandwidth *. 3. in
  let kdf_width_y = stddev_y *. bandwidth *. 3. *. 2. in
  let n_kdf_part_points_x = ((truncate (kdf_width_x /. bin_dx)) / 2) * 2 in
  let n_kdf_part_points_y = ((truncate (kdf_width_y /. bin_dy)) / 2) * 2 in
  Info.set info "kdf_width_x"         (Info.Float32 kdf_width_x);
  Info.set info "kdf_width_y"         (Info.Float32 kdf_width_y);
  Info.set info "n_kdf_part_points_x" (Info.Int     n_kdf_part_points_x);
  Info.set info "n_kdf_part_points_y" (Info.Int     n_kdf_part_points_y);

  let show_parameters _ =
    Printf.printf " stddev_x %f\n" stddev_x;
    Printf.printf " stddev_y %f\n" stddev_y;
    Printf.printf " kdf_width_x %f\n" (Info.float_of info "kdf_width_x");
    Printf.printf " kdf_width_y %f\n" (Info.float_of info "kdf_width_y");
    Printf.printf " x_min %f\n" (Info.float_of info "x_min");
    Printf.printf " x_max %f\n" (Info.float_of info "x_max");
    Printf.printf " y_min %f\n" (Info.float_of info "y_min");
    Printf.printf " y_max %f\n" (Info.float_of info "y_max");
    Printf.printf " n_kdf_part_points_x %f\n" (Info.float_of info "n_kdf_part_points_x");
    Printf.printf " n_kdf_part_points_y %f\n" (Info.float_of info "n_kdf_part_points_y");
    Printf.printf " n_data %f\n" (Info.float_of info "n_data");
    Printf.printf " n_hist_bins/n_pdf_points %d\n" (Info.int_of info "n_hist_bins" / Info.int_of info "n_pdf_points");
  in
  pv_debug t show_parameters;
            
  let compile_options = Pocl.compile_options pocl info "KDE" in
  let program = Pocl.compile_program pocl cl_kernel_source compile_options in

  pv_verbose t (fun _ -> Printf.printf "histogram...%!");
  let histogram_array = gpu_compute_histogram pocl info program sl_array true in
  pv_verbose t (fun _ -> Printf.printf "done\n%!");

  pv_verbose t (fun _ -> Printf.printf "kernel filtering rows...%!");
  let partial_pdf_array = gpu_compute_partial_pdf pocl info program histogram_array in
  pv_verbose t (fun _ -> Printf.printf "done\n%!");

  pv_verbose t (fun _ -> Printf.printf "kernel filtering columns...%!");
  let pdf_array = gpu_compute_full_bivariate_pdf pocl info program partial_pdf_array in
  pv_verbose t (fun _ -> Printf.printf "done\n%!");

  let weighted_total = (ODM.sum' pdf_array) *. bin_dx *. bin_dy in
  ODM.div_scalar pdf_array weighted_total
    
